Move particle serialization into ParticleHandler

gassmoeller opened the pull request:
This moves most of the particle serialization functionality into the ParticleHandler class. Some parts remain in World for now, I will address them in a follow-up PR.

bangerth commented:
There is one failing test. Is this expected?

bangerth commented on source/particle/world.cc:
```diff
@@ -148,125 +143,66 @@ namespace aspect
+//      parallel::distributed::Triangulation<dim> &non_const_triangulation =
+//        const_cast<parallel::distributed::Triangulation<dim> &> (triangulation);
+//
+//#if DEAL_II_VERSION_GTE(9,0,0)
```
do you want to get these commented out lines back in?

bangerth commented on source/particle/world.cc:
```diff
@@ -148,125 +143,66 @@ namespace aspect
-          triangulation.notify_ready_to_unpack(data_offset,callback_function);
 
-          apply_particle_per_cell_bounds();
+//#endif
```
and here?

gassmoeller commented:
> There is one failing test. Is this expected?

No that was a bug, although just a missing update of a cached number. Fixed.
The commented lines sneaked in there unintentionally, I removed them. They do not work yet, and can only be used when dealii/dealii#5150 is merged anyway.

gassmoeller commented:
And it is simpler to read the patch in split mode. I mostly just move functions from World to ParticleHandler (and adjust some minor things that are connected to this move).
